### PR TITLE
Import canonical serialize deserialize

### PR DESCRIPTION
**Pull request by 95DDB**
This PR imports the CanonicalSerialize/Deserialize traits from arkwors and implement/derive it for relevant data structures.
There are few open points (for which we should open corresponding issues when merging):

- We can deprecate the ToBytes/FromBytes traits in favour of CanonicalSerialize/CanonicalDeserialize. The proposal would be to remove them entirely (breaking change), or keep only the base traits with a default implementation calling the CanonicalSerialize/CanonicalDeserialize function (still breaking change, but not as much as before);
- We should merge the Serde serialization/deserialization utilities with the new CanonicalSerialize/CanonicalDeserialize implementations: this will allow to produce easily compressed JSON representations of relevant structs

### CODE REVIEW

**Review by phoinic — APPROVED** [review]

File: algebra/src/curves/tests.rs [hunk 1]
```diff
@@ -280,3 +283,338 @@ pub fn curve_tests<G: ProjectiveCurve>() {
...
+        }
+    }
+
+    for _ in 0..ITERATIONS {
+        let biginteger =
+            <<GroupAffine<P> as AffineCurve>::BaseField as PrimeField>::BigInt::rand(rng);
+        let mut bytes = to_bytes![biginteger].unwrap();
+        let mut g = GroupAffine::<P>::from_random_bytes(&bytes);
+        while g.is_none() {
+            bytes.iter_mut().for_each(|i| *i = i.wrapping_sub(1));
```
Comment (UlrichHaboeck75) [inline algebra/src/curves/tests.rs:550]: Hunting and pecking with decreasing the x-coordinate until success?
Most likely alright, but where do we have that piece of code from?

File: algebra/src/curves/tests.rs [hunk 2]
```diff
@@ -280,3 +283,338 @@ pub fn curve_tests<G: ProjectiveCurve>() {
...
+
+            let mut cursor = Cursor::new(&serialized[..]);
+            let p1 = <GroupAffine<P> as CanonicalDeserialize>::deserialize(&mut cursor).unwrap();
+            let p2 = GroupAffine::<P>::from_random_bytes(&serialized).unwrap();
+            assert_eq!(p1, p2);
+        }
+    }
+}
+
+pub fn sw_jacobian_curve_serialization_test<P: SWModelParameters>() {
```
Comment (UlrichHaboeck75) [inline algebra/src/curves/tests.rs:314]: Couldn't we make a test template for SW curves?

File: algebra/src/fields/tests.rs [hunk 3]
```diff
@@ -201,6 +199,88 @@ fn random_serialization_tests<F: Field, R: Rng>(rng: &mut R) {
...
+                Some(DummyFlags)
+            }
+        }
+
+        use crate::serialize::SerializationError;
+        {
+            let mut serialized = vec![0; buf_size];
+            assert!(if let SerializationError::NotEnoughSpace = a
+                .serialize_with_flags(&mut &mut serialized[..], DummyFlags)
+                .unwrap_err()
```
Comment (UlrichHaboeck75) [inline algebra/src/fields/tests.rs:258]: I don't get the role of ``DummyFlags``. Did we initialize it at all?

File: algebra/src/curves/models/short_weierstrass_jacobian.rs [hunk 4]
```diff
@@ -1,17 +1,10 @@
...
     io::{Read, Result as IoResult, Write, Error as IoError, ErrorKind},
     marker::PhantomData,
 };
-
-use crate::{
-    bytes::{FromBytes, ToBytes},
-    curves::{AffineCurve, ProjectiveCurve},
-    fields::{BitIterator, Field, PrimeField, SquareRootField},
-};
+use crate::{bytes::{FromBytes, ToBytes}, curves::{AffineCurve, ProjectiveCurve, models::SWModelParameters as Parameters}, fields::{BitIterator, Field, PrimeField, SquareRootField}, CanonicalSerialize, SerializationError, CanonicalSerializeWithFlags, CanonicalDeserialize, CanonicalDeserializeWithFlags, UniformRand, SemanticallyValid, Error, FromBytesChecked, BitSerializationError, FromCompressedBits, ToCompressedBits, SWFlags};
```
Comment (UlrichHaboeck75) [inline algebra/src/curves/models/short_weierstrass_jacobian.rs:7]: Why here just in a single line?

File: proof-systems/src/darlin/pcd/simple_marlin.rs [hunk 5]
```diff
@@ -18,12 +15,58 @@ use crate::darlin::{
     },
 };
 use poly_commit::ipa_pc::Commitment;
+use std::ops::Deref;
 use std::marker::PhantomData;
 
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""), Eq(bound = ""), PartialEq(bound = ""))]
+#[derive(CanonicalSerialize, CanonicalDeserialize)]
+pub struct MarlinProof<G: AffineCurve, D: Digest>(pub Proof<G::ScalarField, InnerProductArgPC<G, D>>);
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/pcd/simple_marlin.rs:24]: This again points at moving away from being as generic as possible. 
I like it!

**Review by UlrichHaboeck75 — APPROVED** [review]
Besides that a test template for SW curves would be nice, approved.

**Review by albertog78 — APPROVED** [review]

**Pull request merged**